KIPs/kip-113.md

```diff
@@ -0,0 +1,252 @@
+
+The function emits a `PubkeyRegistered` event.
+
+_Note_ The function is not able to verify the validity of the public key and the proof-of-possession due to the lack of [EIP-2537](https://eips.ethereum.org/EIPS/eip-2537). See [validation](#validation) for off-chain validation.
+
+```solidity
+function registerPublicKey(bytes memory publicKey, bytes memory pop)
+```
+
+##### unregisterPublicKey
+
+Unregisters the public key associated with the given `cnNodeId` address.
+
+The function validates the following requirements:
+
+- The function MUST revert if `cnNodeId` has not been registered.
+- The function SHOULD authenticate and authorize `msg.sender`.
+
+The function emits a `PubkeyUnregistered` event.
+
+```solidity
+function unregisterPublicKey(address cnNodeId)
+```
+
+##### getInfo
+
+Returns the public key and proof-of-possession given a `cnNodeId`.
+
+The function MUST revert if given `cnNodeId` is not registered.
+
+_Note_ The function is not able to verify the validity of the public key and the proof-of-possession due to the lack of [EIP-2537](https://eips.ethereum.org/EIPS/eip-2537). See [validation](#validation) for off-chain validation.
+
+```solidity
+function getInfo(address cnNodeId) external view returns (BlsPublicKeyInfo pubkey)
+```
+
+##### getAllInfo
+
+Returns all the stored node IDs, public keys, and proof-of-possessions at once.
+
+_Note_ The function is not able to verify the validity of the public key and the proof-of-possession due to the lack of [EIP-2537](https://eips.ethereum.org/EIPS/eip-2537). See [validation](#validation) for off-chain validation.
+
+```solidity
+function getAllInfo() external view returns (address[] nodeId, BlsPublicKeyInfo[])
+```
+
+### Core Logic Overview
+
+The `FORK_BLOCK` and `KIP113_CONTRACT_ADDRESS` must be specified per chain in ChainConfig; starting from `FORK_BLOCK`, the node shall read the contract that supports IKIP113 at the address `KIP113_CONTRACT_ADDRESS`. Before `FORK_BLOCK`, the contract should be ready to use, meaning that deploying Registry contract and registering the public keys should be done.
+
+#### Validation
+
+After reading from Registry, the node must perform the following validations:
+
+- [KeyValidate](https://www.ietf.org/archive/id/draft-irtf-cfrg-bls-signature-05.html#section-2.5)
+- [PopVerify](https://www.ietf.org/archive/id/draft-irtf-cfrg-bls-signature-05.html#section-3.3.3)
+
+## Rationale
+
+### Design decisions
+
+#### Registry is not upgradeable
+
+The upgradeable proxy pattern allows the functionality of contracts to be upgraded without changing their addresses.
+However, Registry should not be implemented via upgradeable proxy contract due to its impact on the ecosystem:
+
+- Upgrade can be done covertly by the contract owner.
+- Upgrade can happen at an arbitrary block.
+- Upgrade key can be stolen.
+
+In other words, should Registry be upgraded, Registry must be redeployed to have a different address.
```

Comment:
```suggestion
In other words, if Registry should be upgraded, Registry must be redeployed to have a different address.
```

Reply:
This section is removed because it no longer defines how the registry as in system contract will be deployed. Any developers can deploy them in an upgradeable way.